src/pdn/src/straps.cpp

```diff
@@ -347,6 +348,98 @@ FollowPins::FollowPins(Grid* grid, odb::dbTechLayer* layer, int width)
   }
 }
 
+void FollowPins::findClipCellsObstructions()
+{
+  std::vector<std::shared_ptr<Shape>> power_obs;
+  std::vector<std::shared_ptr<Shape>> ground_obs;
+
+  auto* grid = getGrid();
+  odb::Rect boundary = grid->getGridBoundary();
+  odb::dbTechLayer* layer = getLayer();
+
+  odb::dbNet* power = getDomain()->getPower();
+  odb::dbNet* ground = getDomain()->getGround();
+
+  for (odb::dbInst* inst : grid->getBlock()->getInsts()) {
+    if (!inst->isEndCap()
+        && inst->getMaster()->getType() != odb::dbMasterType::CORE_WELLTAP) {
+      continue;
+    }
+    if (!inst->isFixed() || !boundary.contains(inst->getBBox()->getBox())) {
+      continue;
+    }
+
+    const odb::dbTransform transform = inst->getTransform();
+
+    for (odb::dbITerm* iterm : inst->getITerms()) {
+      bool is_power = iterm->getNet() == power;
+      bool is_ground = iterm->getNet() == ground;
+
+      // Non-power pins might obstruct
+      for (odb::dbMPin* mpin : iterm->getMTerm()->getMPins()) {
+        for (odb::dbBox* box : mpin->getGeometry()) {
+          if (box->getTechLayer() != layer) {
+            continue;
+          }
+          odb::Rect rect = box->getBox();
+          transform.apply(rect);
+
+          auto shape
+              = std::make_shared<Shape>(box->getTechLayer(), rect, Shape::OBS);
+          shape->generateObstruction();
+          if (!is_power) {
+            power_obs.emplace_back(shape);
+          }
+          if (!is_ground) {
+            ground_obs.emplace_back(shape);
+          }
+        }
+      }
+      // Actual obstructions
+      for (odb::dbBox* obstr : inst->getMaster()->getObstructions()) {
+        if (obstr->getTechLayer() != layer) {
+          continue;
+        }
+        odb::Rect rect = obstr->getBox();
+        transform.apply(rect);
+
+        auto shape
+            = std::make_shared<Shape>(obstr->getTechLayer(), rect, Shape::OBS);
+        shape->generateObstruction();
+        power_obs.emplace_back(shape);
+        ground_obs.emplace_back(shape);
+      }
+    }
+  }
+
+  power_obs_tree.clear();
+  power_obs_tree.insert(power_obs.begin(), power_obs.end());
+  ground_obs_tree.clear();
+  ground_obs_tree.insert(ground_obs.begin(), ground_obs.end());
+}
+
+void FollowPins::clipAndAddShape(Shape* shape,
+                                 Shape::ObstructionTree& clip_obstructions)
+{
+  std::vector<Shape*> replacements;
+
+  const int width = getWidth();
+  auto* grid = getGrid();
+
+  if (shape->cut(clip_obstructions, grid, replacements)) {
+    for (Shape* cut_shape : replacements) {
+      if (cut_shape->getRect().dx() > 20 * width) {
+        cut_shape->setLocked();
+        addShape(cut_shape);
+      } else {
```

Comment:
Yeah - because of the two-rings nature, which mimicks our PDK, we get this little unconnected stub.

![image](https://github.com/The-OpenROAD-Project/OpenROAD/assets/78621419/fea2d3d1-e572-41a8-9342-a28e412ae894)